stackable-certs/src/ca/mod.rs

```diff
@@ -294,13 +331,84 @@ where
         let key = ecdsa::SigningKey::new().context(GenerateEcdsaSigningKeySnafu)?;
         self.generate_leaf_certificate(key, name, scope, validity)
     }
+
+    /// Create a [`CertificateAuthority`] from a Kubernetes [`Secret`].
+    #[instrument(name = "create_certificate_authority_from_k8s_secret", skip(secret))]
+    pub fn from_secret(
+        secret: Secret,
+        key_certificate: &str,
+        key_private_key: &str,
+    ) -> Result<Self, SecretError<S::Error>> {
+        if !secret.type_.as_ref().is_some_and(|s| s == TLS_SECRET_TYPE) {
+            return InvalidSecretTypeSnafu.fail();
+        }
+
+        let data = secret.data.as_ref().with_context(|| NoSecretDataSnafu {
+            secret: ObjectRef::from_obj(&secret),
+        })?;
+
+        debug!("retrieving certificate data from secret via key {key_certificate:?}");
+        let certificate_data =
+            data.get(key_certificate)
+                .with_context(|| NoCertificateDataSnafu {
+                    secret: ObjectRef::from_obj(&secret),
+                })?;
+
+        let certificate = x509_cert::Certificate::load_pem_chain(&certificate_data.0)
+            .with_context(|_| ReadChainSnafu {
+                secret: ObjectRef::from_obj(&secret),
+            })?
+            .remove(0);
+
+        debug!("retrieving private key data from secret via key {key_certificate:?}");
+        let private_key_data =
+            data.get(key_private_key)
+                .with_context(|| NoPrivateKeyDataSnafu {
+                    secret: ObjectRef::from_obj(&secret),
+                })?;
+
+        let private_key_data =
+            std::str::from_utf8(&private_key_data.0).context(DecodeUtf8StringSnafu)?;
+
+        let signing_key_pair =
+            S::from_pkcs8_pem(private_key_data).context(DeserializeKeyFromPemSnafu)?;
+
+        Ok(Self {
+            certificate_pair: CertificatePair {
+                key_pair: signing_key_pair,
+                certificate,
+            },
+        })
+    }
+
+    /// Create a [`CertificateAuthority`] from a Kubernetes [`SecretReference`].
+    #[instrument(
+        name = "create_certificate_authority_from_k8s_secret_ref",
+        skip(secret_ref, client)
+    )]
+    pub async fn from_secret_ref(
+        secret_ref: &SecretReference,
+        key_certificate: &str,
+        key_private_key: &str,
+        client: &Client,
+    ) -> Result<Self, SecretError<S::Error>> {
+        let secret_api = client.get_api::<Secret>(&secret_ref.namespace);
+        let secret = secret_api
+            .get(&secret_ref.name)
+            .await
+            .context(GetSecretSnafu {
+                secret_ref: secret_ref.to_owned(),
+            })?;
```

Comment:
**suggestion (non-blocker)**: As discussed, this could/will allocate, so we _should_ use `with_context`.

If it is an `Option` (`snafu::OptionExt` is already imported):

```suggestion
            .with_context(|| GetSecretSnafu {
                secret_ref: secret_ref.to_owned(),
            })?;
```

or if it was a `Result`:

```suggestion
            .with_context(|_| GetSecretSnafu {
                secret_ref: secret_ref.to_owned(),
            })?;
```

Reply:
Fixed in dd51591